compiler/src/dotty/tools/dotc/transform/CheckUnused.scala

```diff
@@ -401,16 +410,27 @@ object CheckUnused:
      * The optional name will be used to target the right import
      * as the same element can be imported with different renaming
      */
-    def registerUsed(sym: Symbol, name: Option[Name], isDerived: Boolean = false)(using Context): Unit =
-      if !isConstructorOfSynth(sym) && !doNotRegister(sym) then
-        if sym.isConstructor && sym.exists then
-          registerUsed(sym.owner, None) // constructor are "implicitly" imported with the class
+    def registerUsed(sym: Symbol, name: Option[Name], notForImport: Boolean = false, isDerived: Boolean = false)(using Context): Unit =
+      if sym.exists && !isConstructorOfSynth(sym) && !doNotRegister(sym) then
+        if sym.isConstructor then
+          registerUsed(sym.owner, None, notForImport) // constructor are "implicitly" imported with the class
         else
-          usedInScope.top += ((sym, sym.isAccessibleAsIdent, name, isDerived))
-          usedInScope.top += ((sym.companionModule, sym.isAccessibleAsIdent, name, isDerived))
-          usedInScope.top += ((sym.companionClass, sym.isAccessibleAsIdent, name, isDerived))
+          // If the symbol is accessible in this scope without an import, do not register it for unused import analysis
+          val notForImport1 =
+            notForImport
+              || (!name.exists(_.toTermName != sym.name.toTermName) && sym.isAccessibleAsIdent)
```

Comment:
Similarly to `isInImport`: can this first part be `name.forall(_.toTermName == sym.name.toTermName)`? (Less negation should make it simpler)